lp2p/host.go

```diff
@@ -0,0 +1,124 @@
+// Package lp2p implements auxiliary functions for go-libp2p integration in CometBFT.
+// The name is chosen to avoid conflicts with the p2p package.
+package lp2p
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/cometbft/cometbft/config"
+	cmcrypto "github.com/cometbft/cometbft/crypto"
+	"github.com/cometbft/cometbft/libs/log"
+	"github.com/libp2p/go-libp2p"
+	"github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/network"
+	"github.com/libp2p/go-libp2p/core/peer"
+	quic "github.com/libp2p/go-libp2p/p2p/transport/quic"
+)
+
+// Host is a wrapper around the libp2p host.
+// Note that host should NOT be responsible for high-level peer management
+// as it's Switch's responsibility.
+type Host struct {
+	host.Host
+	logger      log.Logger
+	configPeers []peer.AddrInfo
+}
+
+// TransportQUIC quic transport.
+// @see https://docs.libp2p.io/concepts/transports/quic
+const TransportQUIC = "quic-v1"
+
+// NewHost Host constructor.
+func NewHost(
+	config *config.P2PConfig,
+	nodeKey cmcrypto.PrivKey,
+	addressBook AddressBookConfig,
+	logger log.Logger,
+) (*Host, error) {
+	if !config.LibP2PEnabled() {
+		return nil, fmt.Errorf("libp2p is disabled")
+	}
+
+	privateKey, err := privateKeyFromCosmosKey(nodeKey)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert private key to libp2p: %w", err)
+	}
+
+	listenAddr, err := AddressToMultiAddr(config.ListenAddress, TransportQUIC)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert %q to multiaddr: %w", config.ListenAddress, err)
+	}
+
+	peers, err := addressBook.DecodePeers()
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode peers from address book: %w", err)
+	}
+
+	// todo: add support for libp2p.ResourceManager() based on p2p.lp2p toml config
+	// todo: add support for libp2p.BandwidthReporter()
+	opts := []libp2p.Option{
+		libp2p.Identity(privateKey),
+		libp2p.ListenAddrs(listenAddr),
+		libp2p.UserAgent("cometbft"),
```

Comment:
Is this standard in some header in existing comet p2p? Potentially we want to include some major version data in this so we can tell which verison of comet is posting different messages in the future?